Use a pipe for builder output on OpenBSD. Reading its PTY master before
the slave opens returns EOF, and a short-lived slave can close before the
startup message is read. A pipe preserves both the startup protocol and
buffered logs until the parent consumes them. Builders on OpenBSD therefore
see a non-terminal stderr.

--- a/unix/build/derivation-builder.cc
+++ b/unix/build/derivation-builder.cc
@@ -116,6 +116,10 @@
      * The process ID of the builder.
      */
     Pid pid;
+
+#ifdef __OpenBSD__
+    AutoCloseFD builderOutPipeWrite;
+#endif
 
     LocalStore & store;
 
@@ -848,6 +852,12 @@
     /* Create the log file. */
     miscMethods->openLogFile();
 
+#ifdef __OpenBSD__
+    Pipe logPipe;
+    logPipe.create();
+    builderOut = std::move(logPipe.readSide);
+    builderOutPipeWrite = std::move(logPipe.writeSide);
+#else
     /* Create a pseudoterminal to get the output of the builder. */
     builderOut = posix_openpt(O_RDWR | O_NOCTTY);
     if (!builderOut)
@@ -870,6 +880,7 @@
 
     if (unlockpt(builderOut.get()))
         throw SysError("unlocking pseudoterminal");
+#endif
 
     buildResult.startTime = time(nullptr);
 
@@ -877,6 +888,9 @@
     startChild();
 
     pid.setSeparatePG(true);
+#ifdef __OpenBSD__
+    builderOutPipeWrite.close();
+#endif
 
     processSandboxSetupMessages();
 
@@ -966,6 +980,11 @@
 
 void DerivationBuilderImpl::openSlave()
 {
+#ifdef __OpenBSD__
+    if (dup2(builderOutPipeWrite.get(), STDERR_FILENO) == -1)
+        throw SysError("cannot pipe standard error into log file");
+    builderOutPipeWrite.close();
+#else
     std::string slaveName = getPtsName(builderOut.get());
 
     AutoCloseFD builderOut = open(slaveName.c_str(), O_RDWR | O_NOCTTY);
@@ -984,6 +1003,7 @@
 
     if (dup2(builderOut.get(), STDERR_FILENO) == -1)
         throw SysError("cannot pipe standard error into log file");
+#endif
 }
 
 #if NIX_WITH_AWS_AUTH
